hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -3054,6 +3052,16 @@ else if ( modelPart instanceof EntityMappingType ) {
 			else {
 				expandToAllColumns = false;
 			}
+
+			final EntityValuedModelPart mapping;
```

Comment:
Unfortunately, I think we will have to add the type as additional parameter to `EntityValuedPathInterpretation.from` because this change would also change the meaning of `where treat(e.association) = e2` to always match the primary key, although it should match the FK parts.